### ISSUE TITLE
Unstoppable ERROR Flood After GridMap Addition Undo

### ISSUE THREAD

**Issue by brainsick** [godotengine/godot#15515]
**Godot version:**
99da466

**OS/device including version:**
Ubuntu 17.10

**Issue description:**
Scene.  Bewildered user sits in front of Godot and wants to make a game.  User creates project and a simple scene, comprising a Camera and a MeshInstance of a Cube.  Success!

Scenario.  User adds GridMap to Camera and is presented with the following errors:

```
ERROR: get_meta: Condition ' !metadata.has(p_name) ' is true. returned: Variant()
   At: core/object.cpp:1025.
ERROR: get: Condition ' !p_rid.is_valid() ' is true. returned: __null
   At: core/rid.h:154.
ERROR: instance_set_visible: Condition ' !instance ' is true.
   At: servers/visual/visual_server_scene.cpp:620.
ERROR: get_meta: Condition ' !metadata.has(p_name) ' is true. returned: Variant()
   At: core/object.cpp:1025.
ERROR: get: Condition ' !p_rid.is_valid() ' is true. returned: __null
   At: core/rid.h:154.
ERROR: instance_set_visible: Condition ' !instance ' is true.
   At: servers/visual/visual_server_scene.cpp:620.
ERROR: get_meta: Condition ' !metadata.has(p_name) ' is true. returned: Variant()
   At: core/object.cpp:1025.
```

User immediately realizes his horrible mistake (not really, that silly user) and undoes it with CTRL+Z.

User is presented with an endless stream of:

```
ERROR: get_global_transform: Condition ' !is_inside_tree() ' is true. returned: Transform()
   At: scene/3d/spatial.cpp:266.
<snip, looping>
```

**Steps to reproduce:**
Be the silly user.

**Minimal reproduction project:**
[Min3DExample.zip](https://github.com/godotengine/godot/files/1613837/Min3DExample.zip)

**Comment by reduz** [comment]
Gridmap module code is here
https://github.com/godotengine/godot/tree/master/modules/gridmap

**Comment by Ivan-Veselov** [comment]
OK, I've been trying to figure this out. Here's what I've found.

After pressing CTRL+Z `GridMapEditor` still stores `GridMap` in its `node` field. According to backtrace editor receives `NOTIFICATION_PROCESS` notification and subsequently calls `get_global_transform()` on the underlying `GridMap`. Because creation of this `GridMap` has been canceled `is_inside_tree()` of `get_global_transform()` returns false which causes one error in log. I guess that `NOTIFICATION_PROCESS` is signaled instantly which creates the flood.

I've started familiarizing myself with Godot just a few days ago, so the following is just my guess.
I saw that several editors subscribe to `"node_removed"` signal and set their underlying `Node` to `NULL` if it has been removed. `GridMapEditor` has similar `_node_removed` method but it's not even binded. I believe that the solution is the folowing: bind `_node_removed` in `_bind_methods()` and add this line somewhere:
```
get_tree()->connect("node_removed", this, "_node_removed");
```
I'm not sure about the last part. I saw that some editors calls `connect` during handling of `NOTIFICATION_ENTER_TREE` others do it when they handle `NOTIFICATION_READY`. `GridMapEditor` already handles `NOTIFICATION_ENTER_TREE`, so I added call to `connect` at the beginning of the corresponding if branch. This removed error flood problem, so I can create a two-line PR. Just not sure that I'm doing everything right.

During my investigation I also noticed that this combination (create node and then immediately CTRL+Z) is handled wrongly by many editors (may be all of them?). After CTRL+Z is pressed created object disappears but the editor is still there. This allows to press one of its menu options, for examples, which immediately leads to a crash. Proposed solution fixes error flood problem but not this one. I'm not sure may be this CTRL+Z problem is a more general one. May be it's a separate problem that requires separate issue.
